eclair-core/src/main/scala/fr/acinq/eclair/channel/Commitments.scala

```diff
@@ -293,100 +295,100 @@ object Commitments {
             }
             val fail = UpdateFailHtlc(commitments.channelId, cmd.id, reason)
             val commitments1 = addLocalProposal(commitments, fail)
-            (commitments1, fail)
-          case Left(_) => throw CannotExtractSharedSecret(commitments.channelId, htlc)
+            Success((commitments1, fail))
+          case Left(_) => Failure(CannotExtractSharedSecret(commitments.channelId, htlc))
         }
-      case None => throw UnknownHtlcId(commitments.channelId, cmd.id)
+      case None => Failure(UnknownHtlcId(commitments.channelId, cmd.id))
     }
 
-  def sendFailMalformed(commitments: Commitments, cmd: CMD_FAIL_MALFORMED_HTLC): (Commitments, UpdateFailMalformedHtlc) = {
+  def sendFailMalformed(commitments: Commitments, cmd: CMD_FAIL_MALFORMED_HTLC): Try[(Commitments, UpdateFailMalformedHtlc)] = {
     // BADONION bit must be set in failure_code
     if ((cmd.failureCode & FailureMessageCodecs.BADONION) == 0) {
-      throw InvalidFailureCode(commitments.channelId)
-    }
-    getHtlcCrossSigned(commitments, IN, cmd.id) match {
-      case Some(htlc) if alreadyProposed(commitments.localChanges.proposed, htlc.id) =>
-        // we have already sent a fail/fulfill for this htlc
-        throw UnknownHtlcId(commitments.channelId, cmd.id)
-      case Some(_) =>
-        val fail = UpdateFailMalformedHtlc(commitments.channelId, cmd.id, cmd.onionHash, cmd.failureCode)
-        val commitments1 = addLocalProposal(commitments, fail)
-        (commitments1, fail)
-      case None => throw UnknownHtlcId(commitments.channelId, cmd.id)
+      Failure(InvalidFailureCode(commitments.channelId))
+    } else {
+      getHtlcCrossSigned(commitments, IN, cmd.id) match {
+        case Some(htlc) if alreadyProposed(commitments.localChanges.proposed, htlc.id) =>
+          // we have already sent a fail/fulfill for this htlc
+          Failure(UnknownHtlcId(commitments.channelId, cmd.id))
+        case Some(_) =>
+          val fail = UpdateFailMalformedHtlc(commitments.channelId, cmd.id, cmd.onionHash, cmd.failureCode)
+          val commitments1 = addLocalProposal(commitments, fail)
+          Success((commitments1, fail))
+        case None => Failure(UnknownHtlcId(commitments.channelId, cmd.id))
+      }
     }
   }
 
-  def receiveFail(commitments: Commitments, fail: UpdateFailHtlc): Either[Commitments, (Commitments, Origin, UpdateAddHtlc)] =
+  def receiveFail(commitments: Commitments, fail: UpdateFailHtlc): Try[(Commitments, Origin, UpdateAddHtlc)] =
     getHtlcCrossSigned(commitments, OUT, fail.id) match {
-      case Some(htlc) => Right((addRemoteProposal(commitments, fail), commitments.originChannels(fail.id), htlc))
-      case None => throw UnknownHtlcId(commitments.channelId, fail.id)
+      case Some(htlc) => Success((addRemoteProposal(commitments, fail), commitments.originChannels(fail.id), htlc))
```

Comment:
Can `commitments.originChannels` not contain `fail.id`?
If that happens this will throw inside a Success, so we're not honoring the signature (still throwing an exception instead of wrapping it inside a `Failure`).
This happens in several places.

Reply:
It looks ok to me in the places where we're doing it, but worth another close look.

Reply:
> Can commitments.originChannels not contain fail.id?

It should, but we could add:

```suggestion
      case Some(htlc) if commitments.originChannels.contains(fail.id) => Success((addRemoteProposal(commitments, fail), commitments.originChannels(fail.id), htlc))
```

Feels a bit fragile tbh, it is really an issue with the model I think.